### PR TITLE
Feature/opt map imp

### PR DESCRIPTION
**Pull request by liamholmes31**
What?
- removed layered to/from native calls for transform
- moved to expression-first approach to avoid materialising data many times

https://github.com/azukds/tubular/issues/448

### CODE REVIEW

**Comment by liamholmes31** [comment]
Some info from the profiling (ran on 3 columns to get some different types/scenarios in):
- original (non narwhals tubular) - ~0.0088s
- pre optimised narwhals tubular + pandas - ~0.037s
- optimised narwhals tubular + pandas - ~0.022s
- optimised narwhals tubular + polars - ~0.0056s

so we are still taking a hit short term when working with pandas, but hopefully this work helps to keep it manageable until we move to polars

File: tubular/mapping.py [hunk 1]
```diff
@@ -89,31 +91,41 @@ def __init__(
 
             # Assign the mapping to the single null key if it exists
             if len(null_keys) != 0:
-                null_mappings[col] = mappings[col][null_keys[0]]
+                mappings_from_null[col] = mappings[col][null_keys[0]]
 
         self.mappings = mappings
-        self.null_mappings = null_mappings
+        self.mappings_from_null = mappings_from_null
```
Comment (cjmwills) [inline tubular/mapping.py:101]: S: can we add this attribute to the docstring? It's not currently mentioned there

File: tubular/mapping.py [hunk 2]
```diff
@@ -77,7 +79,7 @@ def __init__(
             msg = f"{self.classname()}: mappings has no values"
             raise ValueError(msg)
 
-        null_mappings = {col: None for col in mappings}
+        mappings_from_null = {col: None for col in mappings}
         for col in mappings:
```
Comment (cjmwills) [inline tubular/mapping.py:87]: Q: this block is designed to handle cases where the mapping key is null, but the use of beartype validation means that can't ever be possible. Won't it reject null keys from the start?

File: tubular/mapping.py [hunk 3]
```diff
@@ -149,9 +172,87 @@ class attribute, indicates whether transformer has been converted to polars/pand
 
     polars_compatible = True
 
+    def _create_mapping_conditions_and_outcomes(
```
Comment (cjmwills) [inline tubular/mapping.py:180]: S: could this (and `_combine_mappings_into_expression`) be static methods and then we can remove the reference to self? Doesn't seem to be needed

File: tubular/mapping.py [hunk 4]
```diff
@@ -149,9 +172,87 @@ class attribute, indicates whether transformer has been converted to polars/pand
...
+            else (
+                nw.col(col) == key,
+                nw.lit(mappings[col][key], dtype=dtype),
+            )
+        )
+
+    def _combine_mappings_into_expression(
+        self,
+        col: str,
+        conditions_and_outcomes: list[tuple[nw.Expr, nw.Expr]],
```
Comment (cjmwills) [inline tubular/mapping.py]: Q: should the typehint for this be a dict rather than a list?

File: tubular/mapping.py [hunk 5]
```diff
@@ -149,9 +172,87 @@ class attribute, indicates whether transformer has been converted to polars/pand
...
+        Returns
+        -------
+        nw.Expr: prepared mapping expression
+
+        """
+
+        return reduce(
+            lambda expr, condition_and_outcome: nw.when(condition_and_outcome[0])
+            .then(condition_and_outcome[1])
+            .otherwise(expr),
```
Comment (cjmwills) [inline tubular/mapping.py:251]: S: It took me a while to get my head around what was happening here, with narwhals syntax in a lambda function in a reduce function. Potentially my unfamiliarity with narwhals and the abstract nature of narwhals expressions, but some more comments to explain what's happening could be useful (doesn't need to hold up the PR though)

File: tubular/mapping.py [hunk 6]
```diff
@@ -160,113 +261,93 @@ def transform(self, X: DataFrame) -> DataFrame:
...
-        # restore original index for pandas
-        if nw.get_native_namespace(X).__name__ == "pandas":
-            X = nw.to_native(X)
-            X.index = index
-
-        return X[column_order]
+        return _return_narwhals_or_native_dataframe(X, return_native)
 
 
 class MappingTransformer(BaseMappingTransformer, BaseMappingTransformMixin):
```
Comment (cjmwills) [inline tubular/mapping.py:362]: Q: is the reference to the use of pd.Series.replace in the docstring still relevant? If not we should update

File: tests/mapping/test_MappingTransformer.py
Comment (cjmwills) [inline tests/mapping/test_MappingTransformer.py:1]: P: some great tests in here that help ensure lots of edge cases are covered!

**Comment by liamholmes31** [comment]
updated description with correct issue, thanks!

File: tubular/mapping.py (hunk 2 above)
Comment (liamholmes31) [reply tubular/mapping.py:87]: hm, does it not go through this type hint which it would pass?
```
mappings: dict[str, dict[Any, Any]]
```

File: tubular/mapping.py (hunk 1 above)
Comment (liamholmes31) [reply tubular/mapping.py:101]: added here https://github.com/azukds/tubular/pull/458/commits/bc1a648d841a4584311d1c246d7d9fe70bc683c3

File: tubular/mapping.py (hunk 3 above)
Comment (liamholmes31) [reply tubular/mapping.py:180]: good call, updated here https://github.com/azukds/tubular/pull/458/commits/57a41972878601f84559a66d4cb2803c4269d0f2

File: tubular/mapping.py (hunk 2 above)
Comment (liamholmes31) [reply tubular/mapping.py:87]: updated here https://github.com/azukds/tubular/pull/458/commits/a57abc0c02a145ec1394a3fb384a9f92206a524a

File: tubular/mapping.py (hunk 4 above)
Comment (liamholmes31) [reply tubular/mapping.py]: thanks updated here https://github.com/azukds/tubular/pull/458/commits/a57abc0c02a145ec1394a3fb384a9f92206a524a

File: tubular/mapping.py (hunk 5 above)
Comment (liamholmes31) [reply tubular/mapping.py:251]: added an example comment here https://github.com/azukds/tubular/pull/458/commits/a3a21ca63f409a7faa35ed03f5126823a7c77775

File: tubular/mapping.py (hunk 6 above)
Comment (liamholmes31) [reply tubular/mapping.py:362]: thanks updated the docstring here https://github.com/azukds/tubular/pull/458/commits/89c79dfd8c7bdfbdb749c4318eef2837e9fcb5a0

File: tubular/mapping.py (hunk 2 above)
Comment (cjmwills) [reply tubular/mapping.py:87]: you're correct, I remember trying this and getting an error, must've been looking at the wrong thing
